File: CSkyZero_V3/tictactoe/tictactoe_train.cpp
#include <iostream>

#include "../alphazero_parallel.h"
#include "../envs/tictactoe.h"

namespace {

skyzero::AlphaZeroConfig build_tictactoe_config() {
    skyzero::AlphaZeroConfig cfg;
    cfg.board_size = 3;
    cfg.num_blocks = 2;
    cfg.num_channels = 32;
    cfg.lr = 1e-3f;
    cfg.weight_decay = 3e-5f;

    cfg.num_simulations = 8;
    cfg.gumbel_m = 4;
    cfg.gumbel_c_visit = 50.0f;
    cfg.gumbel_c_scale = 1.0f;

    cfg.move_temperature_init = 0.8f;
    cfg.move_temperature_final = 0.2f;

    cfg.batch_size = 128;
    cfg.min_buffer_size = 500;
    cfg.linear_threshold = 2048;
    cfg.replay_alpha = 0.75f;
    cfg.max_buffer_size = 100000;
    cfg.train_steps_per_generation = 50;
    cfg.target_replay_ratio = 5.0f;
    cfg.savetime_interval = 120;
    cfg.file_name = "tictactoe";
    cfg.data_dir = "data/tictactoe";
    cfg.save_on_exit = true;

    cfg.fpu_reduction_max = 0.0f;
    cfg.root_fpu_reduction_max = 0.0f;

    cfg.device = torch::cuda::is_available() ? torch::kCUDA : torch::kCPU;
    return cfg;
}

}  // namespace

int main() {
    std::signal(SIGINT, skyzero::signal_handler);
    auto cfg = build_tictactoe_config();
    skyzero::AlphaZeroParallelConfig pcfg;
    pcfg.num_workers = 40;
    pcfg.num_inference_servers = 1;
    pcfg.inference_batch_size = 128;
    pcfg.inference_batch_wait_us = 30;
    pcfg.leaf_batch_size = 4;
    pcfg.max_games_to_process_per_tick = 200;
    pcfg.idle_sleep_ms = 0;

    skyzero::TicTacToe game;
    auto model = skyzero::ResNet(game.board_size, game.num_planes, cfg.num_blocks, cfg.num_channels);
    model->to(cfg.device);

    torch::optim::AdamW optimizer(
        model->parameters(),
        torch::optim::AdamWOptions(cfg.lr).weight_decay(cfg.weight_decay)
    );

    skyzero::AlphaZeroParallel<skyzero::TicTacToe> az(game, model, optimizer, cfg, pcfg);
    az.load_checkpoint();

    std::cout << "Start TicTacToe training, device=" << (cfg.device.is_cuda() ? "cuda" : "cpu") << "\n";
    az.learn();
    std::cout << "Training finished.\n";
    return 0;
}

File: SkyZero_V0/tictactoe/tictactoe_train.py
import sys
import os
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from nets import ResNet
from envs.tictactoe import TicTacToe
from alphazero import AlphaZero
from alphazero_parallel import AlphaZeroParallel
import numpy as np
import torch.optim as optim

train_args = {

    "num_iterations": 100,  # 总共训练多少个iteration。设置为0即一直跑，CtrlC停止。
    
    "num_blocks": 3,       # 网络block数量
    "num_channels": 32,    # 网络channels
    "lr": 0.001,           # 学习率
    "weight_decay": 3e-5,  # 权重衰减系数

    "num_simulations": 50,      # 模拟次数
    "dirichlet_alpha": 1,     # 控制Dirichlet噪声的锐度，不用动
    "dirichlet_epsilon": 0.25,  # 控制Dirichlet噪声在根节点的占比，不用动

    "batch_size": 128,                # 训练时的batch_size
    "train_steps_per_iteration": 50,  # train阶段训多少个batch
    "target_ReplayRatio": 4,          # 样本回放率，即每个样本会被训练的次数。

    "move_temperature": 1.0,   # 落子温度
    "half_life": 6,            # 每一局经过这个步数之后改为argmax落子

    "min_buffer_size": 6400,        # 最小多少样本开始训（需 >= batch_size * train_steps_per_iteration）
    "max_buffer_size": 1e7,      # 最大容纳多少样本（受内存限制）
    "window_exponent": 0.65,        # 不用动
    "window_expand_per_row": 0.4,   # 不用动
    
    "save_interval": 10,           # 隔几个iteration保存一次checkpoint
    "data_dir": "data/tictactoe_",
    "device": "mps",

    "num_workers": 12,     # 并行worker数量
}

if __name__ == "__main__":
    np.set_printoptions(precision=2, suppress=True)
    game = TicTacToe()
    model = ResNet(game, num_blocks=train_args["num_blocks"], num_channels=train_args["num_channels"])
    optimizer = optim.AdamW(model.parameters(), lr=train_args["lr"], weight_decay=train_args["weight_decay"])

    alphazero = AlphaZeroParallel(game, model, optimizer, train_args)
    alphazero.load_checkpoint()
    alphazero.learn()
